src/core/qgsvectorlayerjoinbuffer.cpp

```diff
@@ -490,3 +517,137 @@ void QgsVectorLayerJoinBuffer::connectJoinedLayer( QgsVectorLayer *vl )
   connect( vl, &QgsVectorLayer::layerModified, this, &QgsVectorLayerJoinBuffer::joinedLayerModified, Qt::UniqueConnection );
   connect( vl, &QgsVectorLayer::willBeDeleted, this, &QgsVectorLayerJoinBuffer::joinedLayerWillBeDeleted, Qt::UniqueConnection );
 }
+
+bool QgsVectorLayerJoinBuffer::addFeature( const QgsFeature &feature ) const
+{
+  return addFeatures( QgsFeatureList() << feature );
+}
+
+bool QgsVectorLayerJoinBuffer::addFeatures( const QgsFeatureList &features ) const
+{
+  if ( !containsJoins() )
+    return false;
+
+  // try to add/update a feature in each joined layer
+  Q_FOREACH ( const QgsVectorLayerJoinInfo &info, vectorJoins() )
+  {
+    QgsVectorLayer *joinLayer = info.joinLayer();
+
+    if ( joinLayer && joinLayer->isEditable() && info.isEditable() && info.hasUpsertOnEdit() )
+    {
+      QgsFeatureList joinFeatures;
+
+      Q_FOREACH ( const QgsFeature &feature, features )
+      {
+        const QgsFeature joinFeature = info.extractJoinedFeature( feature );
+
+        // we don't want to add a new feature in joined layer when the id
+        // column value yet exist, we just want to update the existing one
+        const QVariant idFieldValue = feature.attribute( info.targetFieldName() );
+        const QString filter = QgsExpression::createFieldEqualityExpression( info.joinFieldName(), idFieldValue.toString() );
+
+        QgsFeatureRequest request;
+        request.setFilterExpression( filter );
+        request.setLimit( 1 );
```

Comment:
A possible performance improvement here:

You could set the no geometry flag on the request, and also request no attributes. That will speed up this check a lot. Then if the feature is found, use changeGeometry/changeAttributeValue instead of changeFeature (which also fires off a check for feature existence itself...)

Bonus points: add a `bool featureExists( QgsFeatureId id)` method in QgsVectorLayer to remove the duplicate code.

Reply:
More thoughts: to do a featureExists(...) properly, it should be at the provider level, with a method in QgsVectorLayer which first checks the added/deleted buffers. Providers could bypass a lot of unnecessary feature parsing code if all that's needed is a check that a feature id is present. I'd only override the base class featureiterator approach for the OGR provider though, since that's what will benefit this use case.

Reply:
I set the no geometry flag and replace the *updateFeature* call by *changeAttributeValue* as you suggested. 

However, the aim of the feature request is to get a feature based on a column value (the join field name) thanks to an expression filter and not a feature id. So I don't understand how a *featureExists( bool fid )* method could be useful here. Can you enlighten me?

Reply:
Ah, you're right. I read the code wrong.